pkg/core/storage/memory_store.go

```diff
@@ -23,73 +24,88 @@ type MemoryBatch struct {
 
 // Put implements the Batch interface.
 func (b *MemoryBatch) Put(k, v []byte) {
-	b.MemoryStore.put(string(k), slice.Copy(v))
+	put(b.MemoryStore.chooseMap(k), string(k), slice.Copy(v))
 }
 
 // Delete implements Batch interface.
 func (b *MemoryBatch) Delete(k []byte) {
-	b.MemoryStore.drop(string(k))
+	drop(b.MemoryStore.chooseMap(k), string(k))
 }
 
 // NewMemoryStore creates a new MemoryStore object.
 func NewMemoryStore() *MemoryStore {
 	return &MemoryStore{
-		mem: make(map[string][]byte),
+		mem:  make(map[string][]byte),
+		stor: make(map[string][]byte),
 	}
 }
 
 // Get implements the Store interface.
 func (s *MemoryStore) Get(key []byte) ([]byte, error) {
 	s.mut.RLock()
 	defer s.mut.RUnlock()
-	if val, ok := s.mem[string(key)]; ok && val != nil {
+	m := s.chooseMap(key)
+	if val, ok := m[string(key)]; ok && val != nil {
 		return val, nil
 	}
 	return nil, ErrKeyNotFound
 }
 
+func (s *MemoryStore) chooseMap(key []byte) map[string][]byte {
+	if len(key) > 0 {
+		kp := KeyPrefix(key[0])
+		if kp == STStorage || kp == STTempStorage {
+			return s.stor
+		}
+	}
+	return s.mem
```

Comment:
Maybe we should treat it as an error when zero-lenght key is passed to `chooseMap`, otherwise, it's ambiguous. Currently, the only place where it's important is in `(s *MemCachedStore) seek` and there you check the length of `rng.Prefix` before passing it to `chooseMap`.

Reply:
Frankly, I think we can make Seek only work with non-empty Prefix. The DB scheme we've got has one byte for the type of the data and we never ever search through the whole DB, it just makes no sense, we usually need some type-specific part of it.